snake4.6
.TH SNAKE4 6 "11 October 1996" \" -*- nroff -*-

.SH NAME
snake4, snake4scores \- Game starring a fruit-eating snake.

.SH SYNOPSIS
.B snake4
[\-hHV] [\-M file]
[\-\-help] [\-\-highscores] [\-\-version]
[\-\-merge\-highscores=file]

.B snake4scores

.SH DESCRIPTION
The
.B snake4
program is my fourth implementation of the snake game, written for
Unix and X11. Previous versions were written for MS-DOS.

Move the snake around the screen and eat food. The snake grows in
length as you eat. Avoid hitting the fence, eating lethal mushrooms
and rotten food, and biting your own tail.

Now and then a slim\-pill will bounce around. If you eat this, you will
get shorter, and you will gain bonus points for every length you
loose. The bonus is also given when a round is over.

When food starts blinking, it's about to rot. Rotten food is
poisonous, and thus uneatable. As long as it blinks, it can still be
eaten.

Oh, yes \- I almost forgot: You may be visited by the evil headbanger.
Take my advice and run away from him before he hits you in the head!
He won't stay long.

The keys used to control the game are given at startup.

.B snake4
features a site\-wide highscore file. The file contains the top 200
users. Each user may appear once only in the highscore table.

For technical reasons,
.B snake4
connects to the X11 server before parsing options. This makes it
impossible to eg. browse the highscore table using \-H if no server is
available.
To be able to see the scores in such a case,
.B snake4scores
produces the same output as
.B snake4
with option \-H, but does no attempt on connecting to the X11 server.

.SS OPTIONS
.TP
.I "\-h, \-\-help"
Print a usage message on standard output, and exit successfully.

.TP
.I "\-H, \-\-highscores"
Dump the highscore list to standard output, and exit successfully.
Starting
.B snake4
with the name
.B snake4score
does the same thing, without first connecting to the X11 server.  Note
that this will possibly print more entries than the 20 or so that will
fit within the game startup screen.

.TP
.I "-V, \-\-version"
Print version information on standard output, then exit successfully.

.TP
.I "\-M, \-\-merge\-highscores=file"
Merge separate highscore file
.I file
into the default highscore file, and exit successfully. Displays the
new highscore table. This option is for the game owner or root only.

.SH X11 OPTIONS
In addition to the above mentioned options,
.B snake4
accepts the standard X11 toolkit options.

.SH AUTHOR
Sverre H. Huseby <shh@thathost.com>.


ChangeLog
2014-08-14  Alexandre Dantas  <eu@alexdantas.net>

	* Release: 1.0.14

	* Makefile fix after renaming README to README.md

2014-08-13  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.13

	* Code cleanup by Alexandre Dantas

2002-03-02  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.12

	* All: Updated contact information.

1998-07-05  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.11

	* Updated author information, and changed license to Artistic
	License, to please sunsite.

	* pixmaps/instructions: removed old mail address.

Wed Aug 13 22:39:30 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Release: 1.0.10

	* score.c (scoreWriteScoreLines): Makes sure no empty fields are
	written out, as strtok doesn't like empty tokens.

Mon Aug 11 21:30:35 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Release: 1.0.9

	* Makefile: Added possibility to make a binary distribution.

Wed Mar 19 20:11:25 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Version: 1.0.8

	* win.c (winParseOptions): Added a terminating NULL to the
	resources.

Mon Nov 11 22:04:24 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.7

	* score.c: Use 3 positions for position number on score list.

Tue Oct 15 11:32:10 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.6

	* pixmaps/instructions*: Included score values in the instructions.

Fri Oct 11 11:15:15 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.5

	* score.c (scoreGetThisPlayerIndex): New function.

	* board.c (boardShowHighscores): No longer shows the top 20, but
	rather the top 10, plus a few centered around the current player.

Fri Oct 11 11:15:15 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.4

	* snake.c: Added idling timeout for start of rounds. The game is
	over if the user waits too long (a couple of minutes) before
	moving the snake.

Fri Oct 11 10:31:55 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.3

	* Increased number of entries in the highscore file, and made a
	constant for how many scores may be showed inline.

Thu Oct 10 13:41:53 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.2

        * game.c (gameInit): Fixed bug. Random seed was initialized using
        clock(), but clock returns number of ticks since the last call to
        clock(), so it was always zero.

Wed Oct  9 17:48:49 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.1

	* snake4.c (main): Now shows highscores if called as snake4scores.

	* Makefile: At installation, creates a symlink called snake4scores
	to snake4.

Tue Oct  8 06:56:01 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.0 (uploaded to sunsite.unc.edu)

	* win.c (winTogglePause): Bugfix: Added test for
	gameEndOfRound. Problems occurred if the player paused the game
	after the snake died.

	* board.c (boardHandleEvent): Moved test for pausing legal to
	winTogglePause in win.c

	* suid.c: Added this file and suid.h. Fixes SUID-problems on
	SunOS.

	* Makefile: Creates snake4.score from snake4.score.ifi by keeping
	sverrehu only.

	* Translated TODO to English.

Sun Oct  6 20:38:50 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.5

	* snake4.c (version): Added info on who compiled the program, and
	when.

Sun Oct  6 17:03:50 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.4

	* board.c (boardHandleEvent): Bugfix: Expose events would both
	show instructions and repaint game objects.

Sun Oct  6 15:05:19 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.3

	* slimpill.c: Slimpill comes less often.

	* headbanger.c: Score for each move of the headbanger, and extra
	bonus when it leaves.

	* fruit.c (fruitHandleTick): At least two fruits available.

Sat Oct  5 11:19:52 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version 0.9.2

	* snake4.6: Added a man-page.

	* score.c, snake4.c: Added possibility to merge scores from
 	separate highscore file.

Fri Oct  4 22:14:57 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Bumped version up to 0.9.1 - soon finished!

Tue Sep 24 19:38:47 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Initial version. This program was first implemented in assembly
	for MS-DOS by the same author.
	



snake4.c
/* $Id: snake4.c,v 1.7 2002/03/02 21:10:55 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            snake4.c
 *  MODULE OF       snake4 - game of snake eating fruit
 *
 *  DESCRIPTION     
 *
 *  WRITTEN BY      Sverre H. Huseby <shh@thathost.com>
 *
 **************************************************************************/

#include <stdlib.h>
#include <string.h>

#include <shhmsg.h>
#include <shhopt.h>

#include "win.h"
#include "game.h"
#include "score.h"
#include "suid.h"

/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

static void
version(void)
{
    printf(
      "snake4 " VERSION ", by Sverre H. Huseby "
      "(compiled " COMPILED_DATE " by " COMPILED_BY ")\n"
    );
    exit(0);
}

static void
usage(void)
{
    printf(
      "usage: %s [options]\n"
      "\n"
      "  -h, --help                    display this help and exit\n"
      "  -H, --highscores              show highscores and exit\n"
      "  -M, --merge-highscores=FILE   merge highscores from FILE and exit\n"
      "  -V, --version                 output version information and exit\n"
      "\n"
      "In addition, `normal' X11 options are allowed.\n",
      msgGetName()
    );
    exit(0);
}

static void
highscores(void)
{
    scoreDumpHighscores();
    exit(0);
}

static void
merge(char *file)
{
    scoreMergeScoreFile(file);
    scoreDumpHighscores();
    exit(0);
}



/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

int
main(int argc, char *argv[])
{
    optStruct opt[] = {
      /* short long           type        var/func    special       */
        { 'h', "help",        OPT_FLAG,   usage,      OPT_CALLFUNC },
        { 'H', "highscores",  OPT_FLAG,   highscores, OPT_CALLFUNC },
        { 'M', "merge-highscores",
                              OPT_STRING, merge,      OPT_CALLFUNC },
        { 'V', "version",     OPT_FLAG,   version,    OPT_CALLFUNC },
        { 0, 0, OPT_END, 0, 0 }  /* no more options */
    };

    msgSetName(argv[0]);
    suidInit();

    /* to allow showing highscores before connecting to X11 server */
    if (strcmp(msgGetName(), "snake4scores") == 0)
	highscores(); /* exits the program */

    winParseOptions(&argc, argv);
    optParseOptions(&argc, argv, opt, 0);

    gameInit();

    winMainLoop();

    gameFinish();

    return 0;
}
